contracts/DomainRegistry.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "hardhat/console.sol";
+import "./Ownable.sol";
+
+contract DomainRegistry is Ownable {
+    uint256 public totalDomains;
+    uint256 public totalReservedDomains;
+
+    struct Domain {
+        address controller;
+        uint256 deposit;
+    }
+
+    mapping(string => Domain) private domains;
+
+    event DomainReserved(string indexed domainName, address indexed controller, uint256 deposit);
+    event DepositChanged(string indexed domainName, uint256 newDeposit);
+    event DomainControlTransferred(string indexed domainName, address indexed newController);
+    event DomainReleased(string indexed domainName, address indexed controller, uint256 refundAmount);
+
+    constructor() {}
+
+    function reserveDomain(string memory domainName, uint256 initialDeposit) public payable {
+        require(msg.value >= initialDeposit, "Insufficient initial deposit");
```

Comment:
What if I set `initialDeposit` to 0 so I can pay nothing for reservation and there is no spam protection? I can easily take as many domains as I want?
-2 points